File: 22-locator-strategy/java/src/test/java/LocatorStrategy.java
package src.test.java;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;

/**
 * Created by andrew on 8/22/15.
 */
public class LocatorStrategy {
    WebDriver driver;

    @Before
    public void setUp() throws Exception {
        driver = new FirefoxDriver();

    }

    @Test
    public void locatorStategyCSS() {
        driver.get("http://the-internet.herokuapp.com/download");
        String link = driver.findElement(By.cssSelector("a")).getAttribute("href");
        System.out.println(link);
    }

    @Test
    public void locatorStategyPreciseCSS() {
        driver.get("http://the-internet.herokuapp.com/download");
        String link = driver.findElement(By.cssSelector("#content a")).getAttribute("href");
        System.out.println(link);
    }

    @Test
    public void locatorStategyExactCSS() {
        driver.get("http://the-internet.herokuapp.com/download");
        String link = driver.findElement(By.cssSelector("#content .example a")).getAttribute("href");
        System.out.println(link);
    }

    @After
    public void tearDown() throws Exception {
        driver.quit();
    }
}

File: 22-locator-strategy/kotlin/src/test/kotlin/LocatorStrategy.kt
import org.openqa.selenium.By
import org.openqa.selenium.firefox.FirefoxDriver
import kotlin.test.AfterTest
import kotlin.test.Test

class LocatorStrategy {

    val testUrl = "http://the-internet.herokuapp.com/download"

    private val driver by lazy {
        FirefoxDriver()
    }

    @AfterTest
    fun tearDown() {
        // Quit session (closes browser)
        driver.quit()
    }

    @Test
    fun `locator Strategy CSS`() {
        driver.get(testUrl)
        val link = driver.findElement(By.cssSelector("a")).getAttribute("href")
        println(link)
    }

    @Test
    fun `locator Stategy Precise CSS`() {
        driver.get(testUrl)
        val link = driver.findElement(By.cssSelector("#content a")).getAttribute("href")
        println(link)
    }

    @Test
    fun `locator Strategy Exact CSS`() {
        driver.get(testUrl)
        val link = driver.findElement(By.cssSelector("#content .example a")).getAttribute("href")
        println(link)
    }

    @Test
    fun `locator Strategy XPATH`() {
        driver.get(testUrl)
        val link = driver.findElement(By.xpath("//*[@class='row']//a")).getAttribute("href")
        println(link)
        val linkPrecise = driver.findElement(By.xpath("//*[@id='content']//a")).getAttribute("href")
        println(linkPrecise)
        val linkList = driver.findElements(By.xpath("//*[@class='row']//a"))[0].getAttribute("href")
        println(linkList)
        val linkPreciseList = driver.findElements(By.xpath("//*[@id='content']//a"))[0].getAttribute("href")
        println(linkPreciseList)
    }
}